src/s_audio_pulse.c

```diff
@@ -0,0 +1,171 @@
+    client->buffer = getbytes(client->bufsize);
+
+    client->bufattr.maxlength = -1;
+    client->bufattr.tlength = sizeof(float) * channels * blocksize;
+    client->bufattr.prebuf = -1;
+    client->bufattr.minreq = -1;
+    client->bufattr.fragsize = sizeof(float) * channels * DEFDACBLKSIZE;
+
+    client->client = pa_simple_new(
+        NULL,               // Use the default server.
+        client_name,        // Our application's name.
+        dir,
+        NULL,               // Use the default device.
+        stream_name,        // Description of our stream.
+        &client->spec,      // Our sample format.
+        NULL,               // Use default channel map
+        &client->bufattr,   // buffer attributes
+        &err);
+
+    if (!client->client)
+    {
+        error("Failed to initialize PulseAudio '%s': %s", stream_name, pa_strerror(err));
+    }
+
+    if (!client->buffer || !client->client) {
+        client = free_client(client);
+    }
+    if(client && sys_verbose)
+    {
+        pa_usec_t latency;
+        if ((latency = pa_simple_get_latency(client->client, &err)) == (pa_usec_t) -1) {
+            error("pa_simple_get_latency() failed: %s", pa_strerror(err));
+        } else
+            post("PulseAudio latency '%s':\t%0.0f ms (%d bytes)", stream_name, (float)latency/1000., client->bufattr.tlength);
+    }
+    return client;
+}
+
+int
+pulse_open_audio(int inchans, int outchans, int rate, int blocksize)
+{
+    pulse_sink = make_client(PA_STREAM_PLAYBACK, outchans, rate, blocksize);
+    pulse_source = make_client(PA_STREAM_RECORD, inchans, rate, blocksize);
+
+    return !(pulse_sink || pulse_source);
+}
+
+void pulse_close_audio(void)
+{
+    pulse_sink = free_client(pulse_sink);
+    pulse_source = free_client(pulse_source);
+}
+
+int pulse_send_dacs(void)
+{
+
+    if (!STUFF->st_inchannels && !STUFF->st_outchannels) return (SENDDACS_NO);
+
+    if (STUFF->st_inchannels && pulse_source) {
+        int err = 0;
+        t_sample * fp = STUFF->st_soundin;
+        if(pa_simple_read(pulse_source->client, pulse_source->buffer, pulse_source->bufsize, &err) < 0)
+            error("Pulse Audio: %s", pa_strerror(err));
+        else {
+            const int channels = STUFF->st_inchannels;
+            int chan, frame;
+            float*buf = pulse_source->buffer;
+            for(chan = 0; chan<channels; chan++)
+                for(frame = 0; frame<DEFDACBLKSIZE; frame++)
+                    *fp++ = (t_sample)buf[chan + frame * channels];
+        }
+    }
+
+    if (STUFF->st_outchannels && pulse_sink) {
+        int err = 0;
+        t_sample * fp = STUFF->st_soundout;
+        if (fp) {
+            int chan, frame;
+            float*buf = pulse_sink->buffer;
+            const int channels = STUFF->st_outchannels;
+            for(chan = 0; chan<channels; chan++)
+                for(frame = 0; frame<DEFDACBLKSIZE; frame++)
+                    buf[chan + frame * channels] = (float)*fp++;
+        }
```

Comment:
`t_sample * fp = STUFF->st_soundout;` can be const: `const t_sample * fp = STUFF->st_soundout;`

Please optimize the fragment as shown here:

```c
        const t_sample * fp = STUFF->st_soundout;
        if (fp) {
            const int channels = STUFF->st_outchannels;
            float*buf = pulse_sink->buffer;
            if (channels == 1 && (sizeof(*fp) == sizeof(*buf)))
                memcpy(buf, fp, DEFDACBLKSIZE * sizeof(*fp));
            else {
                int chan;
                const t_sample*lim;
                for(chan = 0; chan<channels; chan++) {
                    buf = pulse_sink->buffer + chan;
                    lim = fp + DEFDACBLKSIZE;
                    for(; fp < lim; buf += channels)
                        *buf = (float)*fp++;
                }
            }
        }
```
or if you want to be "fancy" (but less readable), then you can use loop unrolling with duff's device:
```c
        const t_sample * fp = STUFF->st_soundout;
        if (fp) {
            const int channels = STUFF->st_outchannels;
            float*buf = pulse_sink->buffer;
            if (channels == 1 && (sizeof(*fp) == sizeof(*buf)))
                memcpy(buf, fp, DEFDACBLKSIZE * sizeof(*fp));
            else {
                int chan;
                const t_sample*lim;
                for(chan = 0; chan<channels; chan++) {
                    buf = pulse_sink->buffer + chan;
                    lim = fp + DEFDACBLKSIZE;
                    // optimization: Duff's device
                    switch (DEFDACBLKSIZE % 4) {
                        case 0:
                            do {
                                *buf = (float)*fp++; buf += channels;
                        case 3:
                                *buf = (float)*fp++; buf += channels;
                        case 2:
                                *buf = (float)*fp++; buf += channels;
                        case 1:
                                *buf = (float)*fp++; buf += channels;
                            } while (fp < lim);
                    }
                }
            }
        }
```

Reply:
same as above

Reply:
@user706 To expand: Using Duff's device with a *constant* loop length is entirely pointless. The compiler already nows the loop length at compile time and will unroll appropiately. Also, the compiler will automatically use `memcpy` (or proper SIMD instructions) whenever it can, there's no need to put a special check for length=1, in fact you've just added another branch.

Duff's device and similar optimization techniques were invented in the 80s. C/C++ compilers have gone a *long* way since the last 30-40 years, so take any optimization guide you read with a grain of salt. Test it yourself by compiling with full optimization levels and look at the assembly (godbolt.org is you're friend). We should aim to write readable and maintainable code in the first place and only resort to tricks when we actually notice a bottleneck or can't trust the compiler to do the right thing. I always recommend the following quite entertaining video to get a glimpse at the capabilities of modern compilers: https://www.youtube.com/watch?v=w0sz5WbS5AM

Reply:
> either this optimisation fixes the "ridiculously high CPU usage"

No this is completely unrelated to that.

The main point here is just a very minor point.

And that point is:
I personally would not use the following, in the innermost loop
```c
                for(frame = 0; frame<DEFDACBLKSIZE; frame++)
                    buf[chan + frame * channels] = (float)*fp++;
```
but instead
```
                    for(; fp < lim; buf += channels)
                        *buf = (float)*fp++;
```
for the chance that it might be slightly more efficient (while preserving readability).
This is spelling out invariants more easily for the human reader (and for dumb compilers).

(Forget duff!)


But yes, compilers have come a long way. One would need to benchmark this.
Not much use, at guessing which one is faster: [link](ttps://godbolt.org/e?readOnly=true#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAM1QDsCBlZAQwBtMQBGAFlICsupVs1qhkAUgBMAISnTSAZ0ztkBPHUqZa6AMKpWAVwC2tEJNJb0AGTy1MAOWMAjTMRABmAAykADqgWE6rR6hiZmvv6BdDZ2jkYubl6KypiqQQwEzMQEIcam5koqanQZWQQxDs6uHt4Kmdm5YQX15baV8dVeAJSKqAbEyBwA5FLuWFRtANQAItgAYtMAgjrSVgDSDACSAFrYkwBs3OKei8eLAG6oeOiTwKjUrKjMBJMAVE4GVJykk1SPz28qD5vpNbC9kAgRHZWApOF0zuIAOyyE6TNGTNC0Oqg%2BgYyG0aEKSbidzTPFQ5Swkkoxbo37/F7vT50lkkskfL7Us50zHYv5PRlAuls37ArknbnosHk2g/KjEZhGTDi2nomjECAQkTE0mTTzUmUknRagmUg0muRyLrE5GSlm/EjUBVKnVk/XuaS/Z3K9w6WYLZarDY7bAG%2BWK5UyK12%2B3ojniACs0hNxJkXojbxlhMTZJFDwFXVeQMtUY9MaR0wRiMrEpOl2ut1QAH08EYfMRUOdMOh8wCmV85QzAcCftKTYS4VWaTy6Nix/jCa6s5TOCq6fy%2BxzY2iRRzV2XUejeS8N4KfMLdUD9zSY/ORGuj7OTwzXqxWw%2B0erNfil%2B7PRbfXHM0PUNKMZGtJFp1jLdd0%2BThU3/fEPxZN8jCXIU5BmeYlhWdYtl2ZD10dMNzyNSZUINGDSzJICYXhQ9tzRfsl17AhC2LMDZAPVUd2rKtpiGHpWBAIYEyGUhTCGTxxNQESdBLT0FD6AZMFTdxOHEggROkroegAaxABNvGEoZuHEyTpNIWShnEhQQG8LSpME0g4FgJA0DbPB2DIChNVQTzvJAYAAE53FICZWAIVw7IgJxtPEpxbCyABPESNNIDylXoAB5WhWFSpzSCwIwRGAdh4qKvBiFSNQuzswrMAAD1SAworS8SwWUCq3ycBViGSvQsHa0gCGIVt2p6Gh6CYNgOB4fhBGEUQUAUoQ8CcOzIB6VAfGKLERNslI0g0CBLEaUxvksCo4gSQQ/ACPbzruyI9uuqo3G%2BQoavSFons%2Bo69tKbI3o6D7FF%2B/Q8kEOoyhB264V6fpBi4ISRLEiSKusxqAA59gAWkOW5kGQSZgoAOncSYIFwQgSDUkE9AC1x6eteTOM0%2BLdNIAyjKEESzIxwrrNs%2ByRs51GhkkczMYOsWnK5rtiACDRuCAA%3D)

What does the compiler make of it? One would need to benchmark etc.

Reply:
Let me reiterate: this has no recognizable (!) influence on the exteme CPU loads that I am seeing. (Still I believe mine will be slightly faster in the inner loop. But again: This is very low priority; and not related to the extreme CPU loads I'm experiencing. and also the crackling noise artifacts).

Reply:
> for the chance that it might be slightly more efficient

Why do you think so? Don't guess, look at the assembly and benchmark!

> This is spelling out invariants more easily for the human reader (and for dumb compilers).

Ok, readability is certainly debatable (personally, I still find the first version clearer), but the compiler doesn't care. It perfectly understands what the author of the first version wants to do.

Now let's compile the whole thing with "-O3" and see:

https://godbolt.org/z/BhmtbF

GCC 9.3:

The inner loops are 100% equivalent. The difference is that in your version GCC decides to add an extra branch for the special case of `channels1 == 1`.

Clang 10.0:

Clang decides to do some loop unrolling, but the amount depends on the target architecture. E.g., adding "-march=haswell" will make Clang unroll more aggressively.

---

What's the takeaway? Even different compilers don't fully agree on the optimal solution (there's basically a tradeoff between reduced branching VS increased code size), so how can you? Compilers have deep knowledge of the different CPU architectures, so you can trust them that they do something reasonable with such simple code.

When writing code, express your *intent*, don't try to outsmart the compiler. In 2020 there is no need for any kind of pointer arithmetic tricks*), just use plain array subscripts, the compiler knows how to do the right thing. Only suggest micro-optimizations if you've looked at the assembly for various compilers and either proved that the result is indeed faster. Most potential for optimizations is in algorithms and data structure layout, not in how to write a for-loop.

Finally, the code in question runs *once* during each DSP tick, it is certainly not the bottleneck of your application and any attempt to "optimize" it is just wasted time.

*) you see them a lot in Pd, but don't forget that Pd was written in the 90s.

Reply:
> Let me reiterate: this has no recognizable (!) influence on the exteme CPU loads that I am seeing. 

Of course it doesn't, nobody ever claimed that...

Reply:
Nice. Thanks.